js/classes/entities/midground/particle_effect.js

```diff
@@ -0,0 +1,235 @@
+        this.particles = [];
+        this.lifeTime = lifetime;
+        this.color = color;
+        this.opacity = opacity;
+        this.behavior = behavior;
+        this.center = center;
+        this.size = size;
+        this.removeFromWorld = false;
+
+        for (let i = 0; i < amount; i++) {
+            const newPos = new Vector(center.x + Math.random() * spread - spread / 2,
+                                        center.y + Math.random() * spread - spread / 2);
+
+            this.particles.push({pos: newPos, size: size});
+        }
+    }
+
+
+    /**
+     * Updates the particles until their lifetime is up.
+     */
+    update() {
+
+        for (let i = 0; i < this.particles.length; i++) {
+            const currParticle = this.particles[i];
+
+            // shrink the particle
+            const shrinkAmount = this.size / (this.lifeTime / GAME.clockTick);
+            if (currParticle.size - shrinkAmount > 0.01) {
+                currParticle.size -= shrinkAmount;
+            } else {
+                this.removeFromWorld = true; // once one particle is too small, remove the whole effect
+            }
+
+            // apply the behavior
+            if (this.behavior == ParticleEffect.EXPAND) {
+                const speed = 3;
+                const displacement = Vector.direction(this.center, currParticle.pos);
+                currParticle.pos = Vector.add(currParticle.pos, Vector.multiply(displacement, speed));
+
+            } else if (this.behavior == ParticleEffect.CONTRACT) {
+                const speed = 1;
+                const displacement = Vector.direction(this.center, currParticle.pos);
+                currParticle.pos = Vector.subtract(currParticle.pos, Vector.multiply(displacement, speed));
+
+            } else if (this.behavior == ParticleEffect.RISE) {
+                currParticle.pos.y -= 1;
+
+            } else if (this.behavior == ParticleEffect.FALL) {
+                currParticle.pos.y += 1;
+
+            } else if (this.behavior == ParticleEffect.WIGGLE) {
+                currParticle.pos.x += Math.random() * 2 - 1;
+                currParticle.pos.y += Math.random() * 2 - 1;
+            }
+            // if FREEZE, do nothing
+        }
+    }
+
+    /**
+     * Draws the particles on the canvas.
+     */
+    draw() {
+        for (let i = 0; i < this.particles.length; i++) {
+            CTX.beginPath();
+            const canvasPos = Vector.worldToCanvasSpace(this.particles[i].pos);
+            CTX.ellipse(canvasPos.x, canvasPos.y, 
+                this.particles[i].size, this.particles[i].size, 0, 0, Math.PI * 2);
+            CTX.fillStyle = this.color;
+
+            // make the particles slightly transparent
+            CTX.globalAlpha = this.opacity;
+            CTX.fill();
+            CTX.globalAlpha = 1;
+        }
+    }
+
+    // BEHAVIORS
+
+    /**
+     * A behavior for particles that freeze in place.
+     */
+    static get FREEZE() {
```

Comment:
It may be good to include what these actually are in their name, e.g. `FREEZE_BEHAVIOR`.